app/scripts/services/rdp.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("RdpService", function(Logger) {
+    function findRDPPort (service, rdpPortNumber) {
+      var servicePorts = _.get(service, 'spec.ports');
+      return _.find(servicePorts, {targetPort: rdpPortNumber});
+    }
+
+    return {
+      buildRdp: function (address, port) {
+        return '' +
+          'full address:s:' + address + ':' + port +
+          '\nusername:s:Administrator' +
+          '\nscreen mode id:i:2' + // set 2 for full screen
+          '\nprompt for credentials:i:1' +
+          '\ndesktopwidth:i:0' +
+          '\ndesktopheight:i:0' +
+          '\nauthentication level:i:2' +
+          '\nredirectclipboard:i:1' +
+          '\nsession bpp:i:32' +
+          '\ncompression:i:1' +
+          '\nkeyboardhook:i:2' +
+          '\naudiocapturemode:i:0' +
+          '\nvideoplaybackmode:i:1' +
+          '\nconnection type:i:2' +
+          '\ndisplayconnectionbar:i:1' +
+          '\ndisable wallpaper:i:1' +
+          '\nallow font smoothing:i:1' +
+          '\nallow desktop composition:i:0' +
+          '\ndisable full window drag:i:1' +
+          '\ndisable menu anims:i:1' +
+          '\ndisable themes:i:0' +
+          '\ndisable cursor setting:i:0' +
+          '\nbitmapcachepersistenable:i:1' +
+          '\naudiomode:i:0' +
+          '\nredirectcomports:i:0' +
+          '\nredirectposdevices:i:0' +
+          '\nredirectdirectx:i:1' +
+          '\nautoreconnection enabled:i:1' +
+          '\nnegotiate security layer:i:1' +
+          '\nremoteapplicationmode:i:0' +
+          '\nalternate shell:s:' +
+          '\nshell working directory:s:' +
+          '\ngatewayhostname:s:' +
+          '\ngatewayusagemethod:i:4' +
+          '\ngatewaycredentialssource:i:4' +
+          '\ngatewayprofileusagemethod:i:0' +
+          '\npromptcredentialonce:i:1' +
+          '\nuse redirection server name:i:0' +
+          '\n';
+      },
+
+      fileDownload: function (data, fileName, mimeType) {
+        fileName = fileName || "vm.rdp";
+        mimeType = mimeType || "application/rdp";
+
+        var blob = new Blob([data], { type: mimeType });
+        saveAs(blob, fileName);
+      },
+
+      findRDPPort: findRDPPort,
+
+      getAddressPort: function (service, ovm, rdpPortNumber) {
+        var rdpPortObj = findRDPPort(service, rdpPortNumber);
+        if (!rdpPortObj) {
+          return null;
+        }
+
+        var type = _.get(service, 'spec.type');
+        var port = _.get(rdpPortObj, "port");
+        var address;
+        switch (type) {
```

Comment:
So this switch statement is a bit confusing.  All of the `case`s set the variable `address` as a side effect, but then also do a `Logger.warn` (something bad?), then `return null`.  

`Logger.warn()` makes  me think the user should be alerted to something.  `return null;` short-circults the function, but we still have a `break` statement.

I think this is a bit of `switch` abuse that feels a tad brittle to maintain in the long term.  Is it possible to clarify this syntax in any way?

Reply:
For different `type`, there's different way to get the `address`, so legitimate use of `switch` statement.

For particular `type` there should be corresponding reference available, otherwise it's misconfiguration - so `warn` level is used. There's no need to inform the user explicitly within UI in this case due to other flows. But for later potential analysis, corresponding log message should be provided.

I've refactored the `switch` to just set the `address` or log relevant mesage. The `return` is moved after the `switch`, I hope this helps.